docker/Dockerfile

```diff
@@ -12,43 +12,267 @@
+    && rm -rf /var/lib/apt/lists/*
+RUN amdgpu-install -y --accept-eula --usecase=rocm,hiplibsdk --no-dkms
+
+# Download Kokkos release and build HIP backend
+FROM base-build-rocm as build-kokkos-rocm
+WORKDIR /opt/kokkos-${KOKKOS_VERSION}
+ENV LD_LIBRARY_PATH=/opt/rocm/lib:$LD_LIBRARY_PATH
+# RUN wget --progress=dot:giga https://github.com/kokkos/kokkos/archive/refs/tags/${KOKKOS_VERSION}.tar.gz
+# RUN tar -zxf ${KOKKOS_VERSION}.tar.gz && mv /opt/kokkos-${KOKKOS_VERSION}/kokkos-${KOKKOS_VERSION}/* /opt/kokkos-${KOKKOS_VERSION}
+RUN git clone --branch fix-rtld-deepbind https://github.com/cz4rs/kokkos.git /opt/kokkos-${KOKKOS_VERSION}
+RUN CXX=hipcc cmake -S /opt/kokkos-${KOKKOS_VERSION} -B /opt/kokkos-${KOKKOS_VERSION}/build \
+    -G Ninja \
+    -DCMAKE_BUILD_TYPE=Release \
+    -DCMAKE_INSTALL_PREFIX=/opt/kokkos \
+    -DBUILD_SHARED_LIBS:BOOL=ON \
+    -DBUILD_TESTING:BOOL=OFF \
+    -DKokkos_ENABLE_SERIAL:BOOL=ON \
+    -DKokkos_ENABLE_OPENMP:BOOL=ON \
+    -DCMAKE_CXX_STANDARD=20 \
+    -DKokkos_ENABLE_EXAMPLES:BOOL=OFF \
+    -DKokkos_ENABLE_TESTS:BOOL=OFF \
+    -DKokkos_ENABLE_LIBDL:BOOL=OFF \
+    -DKokkos_ENABLE_HIP:BOOL=ON \
+    -DKokkos_ARCH_${AMD_ARCH}=ON
+RUN cmake --build /opt/kokkos-${KOKKOS_VERSION}/build
+RUN cmake --install /opt/kokkos-${KOKKOS_VERSION}/build
 
-COPY . .
+# Download Lightning release and build lightning-kokkos backend with Kokkos-ROCm
+FROM base-build-rocm AS build-wheel-lightning-kokkos-rocm
+WORKDIR /opt/pennylane-lightning
+COPY --from=build-kokkos-rocm /opt/kokkos /opt/kokkos
+ENV CXX=hipcc
+ENV CMAKE_PREFIX_PATH=/opt/kokkos:$CMAKE_PREFIX_PATH
+RUN wget --progress=dot:giga https://github.com/PennyLaneAI/pennylane-lightning/archive/refs/tags/v${LIGHTNING_VERSION}.tar.gz
+RUN tar -zxf v${LIGHTNING_VERSION}.tar.gz && mv pennylane-lightning-${LIGHTNING_VERSION}/* /opt/pennylane-lightning
+RUN pip uninstall -y pennylane-lightning
+RUN PL_BACKEND=lightning_kokkos python setup.py build_ext
+RUN PL_BACKEND=lightning_kokkos python setup.py bdist_wheel
 
-RUN pip install --no-cache-dir wheel \
-    sphinx \
-    && pip install --no-cache-dir -r requirements-dev.txt \
-    && pip uninstall -y PennyLane_Lightning \
-    && make install \
-    && make test-python \
-    && make test-cpp
+# Install lightning-kokkos HIP backend
+FROM base-runtime AS wheel-lightning-kokkos-rocm
+RUN wget --progress=dot:giga https://repo.radeon.com/amdgpu-install/5.3/ubuntu/jammy/amdgpu-install_5.3.50300-1_all.deb
+RUN apt-get update \
+    && apt-get install --no-install-recommends -y \
+    "linux-headers-$(uname -r)" "linux-modules-extra-$(uname -r)" \
+    ./amdgpu-install_5.3.50300-1_all.deb \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+RUN amdgpu-install -y --accept-eula --usecase=hip --no-dkms
+COPY --from=build-kokkos-rocm /opt/kokkos /opt/kokkos
+COPY --from=build-wheel-lightning-kokkos-rocm /opt/pennylane-lightning/dist/ /
+RUN pip install --force-reinstall --no-cache-dir PennyLane_Lightning*.whl
```

Comment:
Given the size of this file, I wonder if it is possible to split the multi-stage builds across files for each respective architecture and backend/target

Reply:
There is certainly a way, let me check.

Reply:
Seems like everything needs to be in one file for Docker to manage the multi-stage building process by itself. If we want to split the Dockerfile, we'll need a script to deal with it ourselves. I think it's not that bad yet :D